### PR TITLE
Run reflow

### PR DESCRIPTION
**Pull request by TApplencourt**
Update our `reflow.py` copy and run it.

Link: https://github.com/KhronosGroup/Vulkan-Docs/blob/main/scripts/reflow.py and thanks @oddhack  :)

This does what we wanted:
 - Force a maximum line width (default 76 char)
 - Split on period 
 - Delete 2+ contiguous spaces
 
 This is a small PR, so it should be easy to review...

### CODE REVIEW

File: adoc/chapters/acknowledgements.adoc
Comment (gmlueck) [inline adoc/chapters/acknowledgements.adoc:54]: Not sure we want these newlines, but it does not hurt.

File: adoc/chapters/device_compiler.adoc
Comment (gmlueck) [inline adoc/chapters/device_compiler.adoc:139]: This took me a while to track down ...

Asciidoctor does not allow you to split a document reference across two lines like this.  Therefore, you should break the line before the "<<device".  As it is now, this causes the following warning when you build the spec:

```
INFO: possible invalid reference: device
```

File: adoc/chapters/opencl_backend.adoc
Comment (gmlueck) [inline adoc/chapters/opencl_backend.adoc:735]: Same thing here.  The line break should be before "<<kernel".

**Review by gmlueck — COMMENTED** [review]
Wow, this is great!  There are a couple problems I point out below, but I think they can be easily fixed.

To validate that nothing got inadvertently changed, I built the PDF both before and after this change and compared the PDF versions.  There were no unexpected changes.  (There were some expected changes on the title page related to the git version and the build date.)

To people who were not at lunch today, this addresses #472.  Thomas was noting how this could be easily done via a script, so I guess he volunteered!

File: adoc/chapters/device_compiler.adoc
Comment (oddhack) [reply adoc/chapters/device_compiler.adoc:139]: In the specs I am editor for, we use the convention that anchor names are hyphenated, e.g. \<\<kernel-bundle, kernel bundles>>.

**Comment by oddhack** [comment]
A few other notes about this script:

- Requires pretty strict adherence to style guidelines, because it's not running the asciidoctor processing pipeline and does not know about every possible asciidoc construct. When I initially did the SYCL format conversion I followed the Vulkan style guidelines in general, so it's not surprising this basically works for you. The Vulkan style guide contains a lot of stuff that is irrelevant to SYCL but much of the markup / writing / miscellaneous sections may be worth extracting.
- Will not reflow long lines in captions and titles (not supported by asciidoc), nor inside tables (possible, but $TOOCOMPLEX for the minimal upside).

It is pretty robust within the scope of what it's intended for. We use it frequently on Vulkan.

**Comment by oddhack** [comment]
@gmlueck it is an orthogonal issue, but you may want to review the Khronos spec copyright and make sure that's up to date with the latest version Neil has (V10, I think?).

File: adoc/chapters/device_compiler.adoc
Comment (TApplencourt) [reply adoc/chapters/device_compiler.adoc:139]: Good catch! I should have try to build it myself. I did a visual review, was obviously not enough. 

My 2cents: It may be easier for us to change the SYCL spec to use hyphen (the regex should be relatively pretty easy) than to try to patch `reflow.py`. At least for me :)

Comment (oddhack) [reply adoc/chapters/device_compiler.adoc:139]: For this type of thing, you can also do 'wdiff -3' to compare tokens in the HTML outputs with usually good results.

FWIW,

> git grep -E '<<[[:alpha:]:]+ '

turns up only these xrefs:

* 'command group'
* 'control flow'
* 'device copyable'
* 'device function'
* 'kernel bundle'

and AFAICT all of them are glossary terms which already have a corresponding hyphenated anchor.

Comment (TApplencourt) [reply adoc/chapters/device_compiler.adoc:139]: And we also use `<<device-function, Device functions>>`, so I guess it was so incoherence in our spec. Will fix.

Comment (gmlueck) [reply adoc/chapters/device_compiler.adoc:139]: Yes, I agree that we should change these references to use the (hyphenated) anchor name.  I did this in #482.

I'm surprised that Asciidoctor doesn't issue an error when there is a reference that does not have a matching anchor.  I guess the syntax "\<\<foo\>\>" will link to *either* an anchor "[[foo']]" or to a glossary term "foo"?

Comment (oddhack) [reply adoc/chapters/device_compiler.adoc:139]: You get anchors implicitly generated in some circumstances, for example section titles (and probably glossary entries, too, though not sure about that). I prefer having explicit control over this so I use explicit anchors everywhere.

There are varying levels of warning / error behavior selectable through command line options. They probably are not turned all the way up at present.

Comment (oddhack) [reply adoc/chapters/device_compiler.adoc:139]: BTW I think if you do something like '\<\<device
anchor, some text>>' you get a generated '\<a href="device
anchor">' which I guess? is legal HTML? But does not map to 'device-anchor'.

**Review by keryell — APPROVED** [review]
Looks globally like a good idea.
Probably you can rephrase "Keep a maximum line width," in the description to fit the mathematical property you want to express.

File: adoc/chapters/acknowledgements.adoc
Comment (oddhack) [reply adoc/chapters/acknowledgements.adoc:54]: You can add more to the endAbbrev regexp in reflow.py.

Comment (TApplencourt) [reply adoc/chapters/acknowledgements.adoc:54]: @oddhack would you prefer us to `fork` reflow.py or should we move `reflow` in a separate repo so it can be shared between projects?

Comment (oddhack) [reply adoc/chapters/acknowledgements.adoc:54]: Go ahead and make a copy (you'll need a couple of other .py files from the scripts/ directory, see the import statements). I keep relevant scripts loosely synced up between the different repos using them but it's not worth the complexity of separate repos.

**Comment by oddhack** [comment]
> @oddhack do you know why 76 characters in "Force a maximum line width (default 76 char)"? Why not 79?

Historical dating back to OpenGL extension days - use whatever is convenient to you. If you use 999999 you get one sentence/line :-) Just make sure you like the value you start with so you don't generate lots of git churn by changing it later.

**Comment by keryell** [comment]
I mean, if we agree on 80 columns inherited from punch cards which percolated for decades into hardware terminals quite before OpenGL and computer graphics, why 76? You loose 4 characters... :-)

**Comment by TApplencourt** [comment]
@gmlueck Added a CI test.

**Comment by TApplencourt** [comment]
This is ready to merge (if we agree on line width...). Summary of changes:

- Copied Vulkan `reflow.py` and dependencies in our repo + updated abbrev list (thanks again @oddhack)
- Run `reflow` on all `chapters` 
- Added a CI script to verify chapters conformance with reflow  
- Minor: Updated some github receipt version

**Comment by gmlueck** [comment]
What column width are you enforcing now?  Is it 76 or 80?  I agree with @keryell, 80 makes more sense than 76.  From @oddhack's response, it didn't seem like there was any particular reason to choose 76.

**Comment by nliber** [comment]
80 columns makes sense -- if folks are still using ADM-5 terminals!  Even the mighty VT-100 had a 132 column mode.

(Honestly, just pick something...)

**Comment by TApplencourt** [comment]
The default so 76. Agree, that 80 if most popular.  `black` -- the python formater -- did choose [88](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-length)  🤷🏽 

> You probably noticed the peculiar default line length. Black defaults to 88 characters per line, which happens to be 10% over 80. This number was found to produce significantly shorter files than sticking with 80 (the most popular), or even 79 (used by the standard library). In general, [90-ish seems like the wise choice](https://youtu.be/wf-BqAjZb8M?t=260).

If we have already two people pushing for 80, I can change to that. Don't want tobikeshedding it too much :)

**Comment by TApplencourt** [comment]
Change default to 80. Funnily enough, it doesn't change anything :)

**Comment by keryell** [comment]
The reason I propose 79 instead of 80 is that sometimes 80 is a problem when there is a 80-hard limit and some editors or modes display some EOL char at the end, like my Emacs whitespace-mode to show all the weird chars.

**Comment by nliber** [comment]
Doesn't making it more than 72 columns make us incompatible with Fortran?

(Sorry; couldn't resist...)

File: adoc/scripts/vkconventions.py
Comment (ProGTX) [inline adoc/scripts/vkconventions.py:113]: A lot of this is Vulkan specific, do we need this, or even this file?

**Review by ProGTX — APPROVED** [review]
LGTM, though I wonder if we need all of the Vulkan related stuff?

Also I agree with the 80 column limit, it's the most standard.

File: adoc/scripts/vkconventions.py
Comment (TApplencourt) [reply adoc/scripts/vkconventions.py:113]: It depends on your definition of `need` :). This file is "required" because of some import. But indeed, they seem to be related to some Vulkan conversion, so in theory, we can just do some modification and remove it. 

But saying that, I think the maintenance will be easier if we just copy/paste everything without "un-necessary-to-get-it-working" modification.  I agree that we don't need it, but it doesn't seem to hurt. The reformated code seems OK.

To be honest, I don't want to learn the inner architecture of `reflow.py` and neither supports a fancy fork. Nobody have time for that :)

Comment (ProGTX) [reply adoc/scripts/vkconventions.py:113]: Easy maintenance is always good to have :)

Comment (oddhack) [reply adoc/scripts/vkconventions.py:113]: The different APIs using these scripts have different conventions and they are capture in the APIConventions interface. So if you needed to, you could create syclconventions.py with the SYCL-specific things and make apiconventions.py import that, instead of vkconventions.py. However, most of the conventions are used by other scripts which you have no current use for in SYCL, so you may not even need that much - you could try replacing the code in apiconventions.py with just

> from spec_tools.conventions import ConventionsBase as APIConventions

and if that works as desired, remove vkconventions.py.

Comment (TApplencourt) [reply adoc/scripts/vkconventions.py:113]: Tried and got 
```
TypeError: Can't instantiate abstract class ConventionsBase with abstract methods api_name, api_prefix, extension_file_path, file_suffix, generate_structure_type_from_name, nextpoin
ter_member_name, null, structtype_member_name, xml_api_name
```

And indeed, you were "too clean", and used `abc.abstractmethod` on those methods. IMO we can merge as is. In the future, if we need a new `syclConvention` we can implement those methods (and maybe refactor the `apiconventions.py` to get a new ENV variable or something around those line)

File: .github/workflows/CI.yml
Comment (oddhack) [inline .github/workflows/CI.yml:45]: If I'm reading the script right, it's requiring that running reflow does nothing? That seems likely to make essentially every PR fail CI until you condition people to do this whenever they make a spec update. I've found it easier as the Vulkan spec editor to just do this when integrating PRs.

**Comment by oddhack** [comment]
@TApplencourt I made a PR against this branch on your fork that just implements the required SYCL conventions for the script to succeed.

File: .github/workflows/CI.yml
Comment (TApplencourt) [reply .github/workflows/CI.yml:45]: >If I'm reading the script right, it's requiring that running reflow does nothing?

Yes. If I remember correctly, it was an ask from @gmlueck .  

It looks like GitHub support issuing [warning](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-a-warning-message) if we want to go this way.

File: adoc/scripts/reflow.py
Comment (gmlueck) [inline adoc/scripts/reflow.py:558]: It looks like commit aa7f7a41187765c6603e8569a228df0c7da810f4 did not actual change the default margin width.  I thin these two lines above are what we need to change.

**Review by gmlueck — COMMENTED** [review]
> The reason I propose 79 instead of 80 is that sometimes 80 is a problem when there is a 80-hard limit and some editors or modes display some EOL char at the end, like my Emacs whitespace-mode to show all the weird chars.

If 79 makes life easier for Emacs users, we can use that.  This PR hasn't changed the column width yet, so it's still easy to choose a different number.  Let's try to decide today, though.

@TApplencourt, there are a number of other PRs that are ready to merge, so I propose the following order:

* We merge whatever PRs the WG approves in the meeting today.
* You merge latest SYCL-2020/master branch into this PR.
* Run reflow with whatever margin limit we decide on (79 or 80).
* I'll merge this PR as an editorial change before the next WG meeting.

File: adoc/scripts/reflow.py
Comment (TApplencourt) [reply adoc/scripts/reflow.py:558]: Oh, good catch. I will just do a big grep on `76`.

**Comment by gmlueck** [comment]
@TApplencourt, other PRs have all been merged now, so we are ready for you to merge the latest "SYCL-2020/master" into this PR, resolve conflicts, and rerun reflow.

I suspect the merge conflicts might be substantial.  If it is easier for you, you could create a new PR with the reflow scripts and run reflow in that new PR.

However you do it, I'll do a final check of the PDF to make sure nothing gets accidentally changed before merging.

**Comment by TApplencourt** [comment]
Please see  #495

**Pull request closed without merging**
